File: 2020/day/23/Main.hs
{-# LANGUAGE ImportQualifiedPost #-}
import Data.Char
import Data.Foldable
import Data.Sequence (Seq(..), (<|), (|>))
import Data.Sequence qualified as Seq

input = "219748365"

newtype Cups = Cups { getLabels :: Seq Int }

instance Show Cups where
    show (Cups cups) = let (xs, _ :<| ys) = Seq.breakl (== 1) cups
                       in map intToDigit (toList (ys <> xs))

step (Cups (curr :<| a :<| b :<| c :<| rest)) =
    let (xs, _dest :<| ys) = Seq.breakl (== dest) (curr <| rest)
        (xs', _curr :<| next :<| ys') = Seq.breakl (== curr) (xs <> (dest <| a <| b <| c <| ys))
    in Cups $ (next <| ys') <> (xs' |> curr)
  where
    dest = let k = until (`notElem` [a, b, c]) pred (curr - 1)
           in if k `elem` rest then k else maximum rest

main = do
    let labels = Seq.fromList (map digitToInt testinput)

    -- Part 1
    print $ iterate step (Cups labels) !! 100

File: 2020/day/23/main.cc
#include <algorithm>
#include <iostream>
#include <string>
#include <vector>

int step(int step, int curr, int min, int max, std::vector<int> &next, bool debug = false) {
    if (debug) std::cerr << "-- move " << step << " --" << std::endl;
    int a = next[curr];
    int b = next[a];
    int c = next[b];

    int dest = curr - 1;

    while (dest == a || dest == b || dest == c) {
        dest--;
    }

    if (dest < min) {
        dest = max;

        while (dest == a || dest == b || dest == c) {
            dest--;
        }
    }

    if (debug) std::cerr << "pick up: " << a << ", " << b << ", " << c << std::endl;
    if (debug) std::cerr << "destination: " << dest << std::endl;

    next[curr] = next[c];
    int tmp = next[dest];
    next[dest] = a;
    next[c] = tmp;
    return next[curr];
}

int main() {
    std::string input("389125467");
    std::vector<int> next(10 + 1);

    for (int i = 0; i < input.length(); i++) {
        next[input[i] - '0'] = input[(i + 1) % 10] - '0';
    }

    next[input[input.length() - 1] - '0'] = input[0] - '0';
    int curr = input[0] - '0';
    int min = 1, max = 9;

    for (int i = 0; i < 100; i++) {
        curr = step(i + 1, curr, min, max, next);
    }

    int x = next[1];

    while (x != 1) {
        std::cout << x;
        x = next[x];
    }

    std::cout << std::endl;

    // Part 2
    max = 1000000;
    next.resize(max + 1);

    // Reset labels
    for (int i = 0; i < input.length(); i++) {
        next[input[i] - '0'] = input[(i + 1) % 10] - '0';
    }

    // Extend with a million elements
    next[input[input.length() - 1] - '0'] = 10;

    for (int i = 10; i < max; i++) {
        next[i] = i + 1;
    }

    next[max] = input[0] - '0';
    curr = input[0] - '0';

    // Do the D.A.N.C.E.
    for (int i = 0; i < 10000000; i++) {
        curr = step(i + 1, curr, min, max, next, false);
    }

    std::cout << ((int64_t) next[1]) * ((int64_t) next[next[1]]) << std::endl;
}